### PR TITLE
Parse safetensors metadata

### PR DESCRIPTION
**Pull request by Wauplin**
Related to https://github.com/huggingface/huggingface_hub/issues/1832 cc @LysandreJik @Narsil 

There are 2 methods:
- `parse_safetensors_file_metadata` => parses a safetensors file on the Hub => the "real" method that parses safetensors
- `get_safetensors_metadata` => takes a repo and parses all safetensors files (if sharded with a `model.safetensors.index.json` file) => more focused towards `transformers` architecture (i.e. opinionated)

I tried to follow more or less the [typescript implementation](https://github.com/huggingface/huggingface.js/blob/main/packages/hub/src/lib/parse-safetensors-metadata.ts) with similar error handling.

### CODE REVIEW

File: src/huggingface_hub/__init__.py [hunk 1]
```diff
@@ -206,6 +207,7 @@
         "merge_pull_request",
         "model_info",
         "move_repo",
+        "parse_safetensors_metadata",
```
Comment (julien-c) [inline src/huggingface_hub/__init__.py:210]:
```suggestion
        "parse_safetensors_file_metadata",
```
maybe?

File: src/huggingface_hub/hf_api.py [hunk 2]
```diff
@@ -4967,6 +4975,227 @@ def snapshot_download(
...
+        """
+        Parse metadata for a safetensors repo on the Hub.
+
+        We first check if the repo has a single safetensors file or a sharded safetensors repo. If it's a single
+        safetensors file, we parse the metadata from this file. If it's a sharded safetensors repo, we parse the
+        metadata from the index file and then parse the metadata from each shard.
+
+        To parse metadata from a single safetensors file, use [`get_safetensors_metadata`].
+
+        For more details regarding the safetensors format, check out https://huggingface.co/docs/safetensors/index#format.
```
Comment (julien-c) [inline src/huggingface_hub/hf_api.py:4995]: you'll be able to hyper-link to this implem from the safetensors doc too BTW (like i had done for the JS implem)

Comment (julien-c) [reply src/huggingface_hub/hf_api.py:4995]: more generally let's always make sure we cross-link stuff as much as possible

**Review by julien-c — APPROVED** [review]

File: src/huggingface_hub/hf_api.py [hunk 3]
```diff
@@ -4967,6 +4975,227 @@ def snapshot_download(
...
+            SafetensorsRepoMetadata(metadata={'total_size': 352494542848}, sharded=True, weight_map={...}, files_metadata={...})
+            >>> len(metadata.files_metadata)
+            72  # All safetensors files have been fetched
+
+            # Parse repo with sharded model
+            >>> get_safetensors_metadata("runwayml/stable-diffusion-v1-5")
+            NotASafetensorsRepoError: 'runwayml/stable-diffusion-v1-5' is not a safetensors repo. Couldn't find 'model.safetensors.index.json' or 'model.safetensors' files.
+            ```
+        """
+        filenames = self.list_repo_files(repo_id=repo_id, repo_type=repo_type, revision=revision, token=token)
```
Comment (LysandreJik) [inline src/huggingface_hub/hf_api.py:5048]: I think the check can be easier here, you mentioned yesterday that `file_exists` is a single HEAD call whereas `list_repo_files` lists the repo:

```
sharded = api.file_exists(repo_id, "model.safetensors.index.json", revision=revision, token=token)
```

as you seem to be using filenames only for this check

File: src/huggingface_hub/hf_api.py [hunk 4]
```diff
@@ -4967,6 +4975,227 @@ def snapshot_download(
...
+            - [`NotASafetensorsRepoError`]: if the repo is not a safetensors repo i.e. doesn't have either a
+              `model.safetensors` or a `model.safetensors.index.json` file.
+            - [`SafetensorsParsingError`]: if a safetensors file header couldn't be parsed correctly.
+        """
+        url = hf_hub_url(
+            repo_id=repo_id, filename=filename, repo_type=repo_type, revision=revision, endpoint=self.endpoint
+        )
+        _headers = self._build_hf_headers(token=token)
+
+        # 1. Fetch first 100kb
```
Comment (LysandreJik) [inline src/huggingface_hub/hf_api.py:5145]: Is this true? If I'm not mistaken @Narsil had told me 1MB but I didn't try it firsthand

**Review by LysandreJik — APPROVED** [review]
Thanks @Wauplin ! API looks great and works great.

File: src/huggingface_hub/hf_api.py (hunk 3 above)
Comment (Wauplin) [reply src/huggingface_hub/hf_api.py]: Yes you're completely right! Thanks for reminding me about it :smile: I switched to `file_exists` in https://github.com/huggingface/huggingface_hub/pull/1855/commits/a4f4d022d840b3bb6849c394a9f48cd518ff2e89.

File: src/huggingface_hub/hf_api.py (hunk 4 above)
Comment (Wauplin) [reply src/huggingface_hub/hf_api.py:5151]: Narsil mentioned me 100kb as a good starting point.

Out of curiosity I've made a quick empirical study. I parsed 2700 files from the top 1000 models tagged as safetensors-compatible on the Hub, sorted by downloads. Out of 2700 files,
- maximum metadata header is 365kb
- 3.2% have a metadata header >=100kb
- 4.1% have a metadata header >=75kb
- 7.5% have a metadata header >= 50kb
- 18% have a metadata header >= 25kb

Given these numbers, `100kb` looks like a good threshold. We could even lower it to 75kb but I don't think it's worth it.

**Comment by Wauplin** [comment]
Thanks for the reviews @julien-c and @LysandreJik!

I'll keep in mind to update the hub docs once this is released (https://github.com/huggingface/huggingface_hub/pull/1855#discussion_r1403309176)

**Pull request merged**